synapse/handlers/federation.py

```diff
@@ -2711,9 +2711,11 @@ async def _update_auth_events_and_context_for_auth(
                             event.event_id,
                             e.event_id,
                         )
-                        context = await self.state_handler.compute_event_context(e)
+                        missing_auth_event_context = (
+                            await self.state_handler.compute_event_context(e)
+                        )
                         await self._auth_and_persist_event(
-                            origin, e, context, auth_events=auth
+                            origin, e, missing_auth_event_context, auth_events=auth
```

Comment:
Main fix is here ^

First stumbled upon in https://github.com/matrix-org/synapse/pull/10245#discussion_r669172325

Reproducible on a federated homeserver when there is a membership auth event as a floating outlier. Then when we try to backfill one of that persons messages, it has missing membership auth to fetch which caused us to [mistakenly replace the `context`](https://github.com/matrix-org/synapse/blob/b181dc402d1e2727375e61931f12969dd04c746f/synapse/handlers/federation.py#L2717) for the message with that of the floating membership `outlier` event. Since `outliers` have no `state` or `state_group`, the error bubbles up when we continue down the persisting route: `sqlite3.IntegrityError: NOT NULL constraint failed: event_to_state_groups.state_group`

Call stack:

 - Bug spot: [`synapse/handlers/federation.py#L2717`](https://github.com/matrix-org/synapse/blob/b181dc402d1e2727375e61931f12969dd04c746f/synapse/handlers/federation.py#L2717)
 - It returns early at [`synapse/handlers/federation.py#L2744`](https://github.com/matrix-org/synapse/blob/b181dc402d1e2727375e61931f12969dd04c746f/synapse/handlers/federation.py#L2744) because it doesn't have any `different_auth`

```
backfill
_auth_and_persist_event
_check_event_auth
# Modifies the context (where the bug is)
_update_auth_events_and_context_for_auth
```

We touch some of the same code on the the receive PDU route but if it is able to reach [this line, it is able to regenerate the `context`](https://github.com/matrix-org/synapse/blob/b181dc402d1e2727375e61931f12969dd04c746f/synapse/handlers/federation.py#L2796) and recover. The cases where we return before and it can reproduce is when it's another [`outlier`, no `different_auth`, or `auth_event` is from another room](https://github.com/matrix-org/synapse/blob/b181dc402d1e2727375e61931f12969dd04c746f/synapse/handlers/federation.py#L2730-L2770).
```
on_receive_pdu
_process_received_pdu
_auth_and_persist_event
_check_event_auth
# Modifies the context (where the bug is)
_update_auth_events_and_context_for_auth
```

Reply:
The main fix disappeared from the diff here because it was merged alongside https://github.com/matrix-org/synapse/pull/10245

This PR is still relevant to add the regression test for it though.